sphinx/application.py

```diff
@@ -123,10 +124,24 @@ def __init__(self, srcdir, confdir, outdir, doctreedir, buildername,
         self.enumerable_nodes = {}              # type: Dict[nodes.Node, Tuple[unicode, Callable]]  # NOQA
         self.html_themes = {}                   # type: Dict[unicode, unicode]
 
-        self.srcdir = srcdir
+        # validate provided directories
+        self.srcdir = abspath(srcdir)
+        self.outdir = abspath(outdir)
+        self.doctreedir = abspath(doctreedir or path.join(outdir, '.doctrees'))
```

Comment:
IMO, it is better that `Sphinx` class simply uses given `doctreedir` argument.

Reply:
OK. I think my reasoning for doing this was that the pattern was duplicated in two places. That's probably not a bad thing though. Perhaps I should raise an exception if these are unset?

Reply:
Yes, I think so. I think new responsibility of `Sphinx` class is regulating paths and checking their existence (validation). So exceptions should be raised if invalid parameter given.

I think reducing code duplication is not important. IMO, it is more important to clear responsibility of each modules. Before, `Sphinx` class and its clients are strongly coupled. `Sphinx` always trusts arguments from clients and does not check them.
With this change, `Sphinx` class becomes defensive. It also means less coupled.

Note: I think choosing `doctreedir` is a responsibility of clients. It's simple.

Reply:
Done